// db/users.js
var fs = require('fs');
var User = require('./User');

var path = __dirname + '/db.json';

var users;

/**
 * DB of users
 */
exports.connect = function() {
    users = require('./db.json');
};

exports.getUser = function(login) {
    return (!contain(login)) ? null : contain(login);
};

exports.addUser = function(login, password) { //may contain some another fields
    if (contain(login))
        return null;
    else {
        var db = fs.readFileSync(path, 'utf-8');
        db = JSON.parse(db);
        db.Users.push(new User(login, password));
        db = JSON.stringify(db, '', 4);
        fs.writeFile(path, db, 'utf-8', function(err) {
            if (err)
                throw new Error(err);
        });
    }
};


exports.deleteUser = function(login) {
    if (!contain(login))
        return null;
    else {
        var db = fs.readFileSync(path, 'utf-8');
        db = JSON.parse(db);
        db.Users.splice(db.Users.indexOf(contain(login)), 1);
        db = JSON.stringify(db, '', 4);
        fs.writeFile(path, db, 'utf-8', function(err) {
            if (err)
                throw new Error(err);
        });
    }
};

function contain(login) {
    var db = fs.readFileSync(path, 'utf-8');
    db = JSON.parse(db);
    for (var i = 0; i < db.Users.length; i++)
        if (db.Users[i].login == login)
            return db.Users[i];
    return false;
}

exports.deleteAll = function() {
    var n = {
        Users: []
    };
    fs.writeFile(path, JSON.stringify(n, '', 4), 'utf-8', function(err) {
        if (err)
            throw new Error(err);
    });
};
